# Data Mining 
### EX 7 Association Mining

#### 8.1 Analyzing the Shopping Basket Data Set

##### 8.1.1 The Shopping Basket data set is provided as an Excel file in ILIAS. Load the data set with the read_excel function.

In [1]:
import pandas as pd

# Read the Excel file into a DataFrame
df = pd.read_excel('ShoppingBaskets.xls')

# Drop the 'BasketNo' column from the DataFrame
df = df.drop('BasketNo', axis=1)

# Display the first few rows of the DataFrame
display(df.head(10))

/var/folders/yj/2fr85zvn2sbbpv3jsq4y3kwh0000gn/T/ipykernel_1357/865161996.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


ThinkPad X220   Asus EeePC  HP Laserjet P2055  2 GB DDR3 RAM  \
0               1           0                  0              0   
1               0           1                  0              1   
2               1           0                  1              0   
3               0           1                  0              1   
4               0           1                  1              1   
5               1           0                  0              0   
6               0           1                  0              1   
7               1           0                  1              0   
8               0           1                  0              0   
9               0           1                  0              1   

   8 GB DDR3 RAM  Lenovo Tablet Sleeve  Netbook-Schutzhülle   HP CE50 Toner  \
0              1                     1                     0              0   
1              0                     0                     1              0   
2              0                     1                     0              1   
3              0                     0                     1              0   
4              0                     0                     0              1   
5              1                     0                     0              0   
6              0                     0                     1              0   
7              0                     1                     0              1   
8              0                     1                     0              0   
9              0                     0                     1              0   

   LT Laser Maus  LT Minimaus  
0              0            1  
1              1            0  
2              1            0  
3              0            1  
4              0            0  
5              1            0  
6              1            1  
7              0            0  
8              0            1  
9              0            1

##### 8.1.2 Mine frequent item sets from the data set using the frequent_itemsets function (support =0.2). Which items are usually bought together with the laptop, the netbook and the printer?

In [2]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets = apriori(df, min_support=0.2, use_colnames= True)
frequent_itemsets 

/Users/sudhirjoon/Library/Python/3.9/lib/python/site-packages/mlxtend/frequent_patterns/fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
0       0.4                                   (ThinkPad X220 )
1       0.6                                       (Asus EeePC)
2       0.3                                (HP Laserjet P2055)
3       0.5                                    (2 GB DDR3 RAM)
4       0.2                                    (8 GB DDR3 RAM)
5       0.4                             (Lenovo Tablet Sleeve)
6       0.4                             (Netbook-Schutzhülle )
7       0.3                                    (HP CE50 Toner)
8       0.4                                    (LT Laser Maus)
9       0.5                                      (LT Minimaus)
10      0.2                (HP Laserjet P2055, ThinkPad X220 )
11      0.2                    (8 GB DDR3 RAM, ThinkPad X220 )
12      0.3             (Lenovo Tablet Sleeve, ThinkPad X220 )
13      0.2                    (HP CE50 Toner, ThinkPad X220 )
14      0.2                    (LT Laser Maus, ThinkPad X220 )
15      0.5                        (Asus EeePC, 2 GB DDR3 RAM)
16      0.4                 (Asus EeePC, Netbook-Schutzhülle )
17      0.2                        (Asus EeePC, LT Laser Maus)
18      0.4                          (Asus EeePC, LT Minimaus)
19      0.2          (HP Laserjet P2055, Lenovo Tablet Sleeve)
20      0.3                 (HP Laserjet P2055, HP CE50 Toner)
21      0.4              (2 GB DDR3 RAM, Netbook-Schutzhülle )
22      0.2                     (LT Laser Maus, 2 GB DDR3 RAM)
23      0.3                       (LT Minimaus, 2 GB DDR3 RAM)
24      0.2              (HP CE50 Toner, Lenovo Tablet Sleeve)
25      0.2                (Lenovo Tablet Sleeve, LT Minimaus)
26      0.2              (LT Laser Maus, Netbook-Schutzhülle )
27      0.3                (LT Minimaus, Netbook-Schutzhülle )
28      0.2  (HP Laserjet P2055, Lenovo Tablet Sleeve, Thin...
29      0.2  (HP Laserjet P2055, HP CE50 Toner, ThinkPad X2...
30      0.2  (HP CE50 Toner, Lenovo Tablet Sleeve, ThinkPad...
31      0.4  (Asus EeePC, 2 GB DDR3 RAM, Netbook-Schutzhülle )
32      0.2         (Asus EeePC, LT Laser Maus, 2 GB DDR3 RAM)
33      0.3           (Asus EeePC, LT Minimaus, 2 GB DDR3 RAM)
34      0.2  (Asus EeePC, LT Laser Maus, Netbook-Schutzhülle )
35      0.3    (Asus EeePC, LT Minimaus, Netbook-Schutzhülle )
36      0.2  (HP Laserjet P2055, HP CE50 Toner, Lenovo Tabl...
37      0.2  (LT Laser Maus, 2 GB DDR3 RAM, Netbook-Schutzh...
38      0.3  (LT Minimaus, 2 GB DDR3 RAM, Netbook-Schutzhül...
39      0.2  (HP Laserjet P2055, HP CE50 Toner, Lenovo Tabl...
40      0.2  (Asus EeePC, LT Laser Maus, 2 GB DDR3 RAM, Net...
41      0.3  (Asus EeePC, LT Minimaus, 2 GB DDR3 RAM, Netbo...

In [3]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.7)
rules

antecedents  \
0                         (8 GB DDR3 RAM)   
1                  (Lenovo Tablet Sleeve)   
2                        (ThinkPad X220 )   
3                            (Asus EeePC)   
4                         (2 GB DDR3 RAM)   
..                                    ...   
60     (Asus EeePC, Netbook-Schutzhülle )   
61           (LT Minimaus, 2 GB DDR3 RAM)   
62    (LT Minimaus, Netbook-Schutzhülle )   
63  (2 GB DDR3 RAM, Netbook-Schutzhülle )   
64                 (Netbook-Schutzhülle )   

                                 consequents  antecedent support  \
0                           (ThinkPad X220 )                 0.2   
1                           (ThinkPad X220 )                 0.4   
2                     (Lenovo Tablet Sleeve)                 0.4   
3                            (2 GB DDR3 RAM)                 0.6   
4                               (Asus EeePC)                 0.5   
..                                       ...                 ...   
60              (LT Minimaus, 2 GB DDR3 RAM)                 0.4   
61        (Asus EeePC, Netbook-Schutzhülle )                 0.3   
62               (Asus EeePC, 2 GB DDR3 RAM)                 0.3   
63                 (Asus EeePC, LT Minimaus)                 0.4   
64  (Asus EeePC, LT Minimaus, 2 GB DDR3 RAM)                 0.4   

    consequent support  support  confidence      lift  leverage  conviction  \
0                  0.4      0.2    1.000000  2.500000      0.12         inf   
1                  0.4      0.3    0.750000  1.875000      0.14         2.4   
2                  0.4      0.3    0.750000  1.875000      0.14         2.4   
3                  0.5      0.5    0.833333  1.666667      0.20         3.0   
4                  0.6      0.5    1.000000  1.666667      0.20         inf   
..                 ...      ...         ...       ...       ...         ...   
60                 0.3      0.3    0.750000  2.500000      0.18         2.8   
61                 0.4      0.3    1.000000  2.500000      0.18         inf   
62                 0.5      0.3    1.000000  2.000000      0.15         inf   
63                 0.4      0.3    0.750000  1.875000      0.14         2.4   
64                 0.3      0.3    0.750000  2.500000      0.18         2.8   

    zhangs_metric  
0        0.750000  
1        0.777778  
2        0.777778  
3        1.000000  
4        0.800000  
..            ...  
60       1.000000  
61       0.857143  
62       0.714286  
63       0.777778  
64       1.000000  

[65 rows x 10 columns]

In [4]:
laptop_rules = rules[rules['antecedents'].astype(str).str.contains('ThinkPad X220')]
printer_rules = rules[rules['antecedents'].astype(str).str.contains('HP Laserjet P2055')]
Netbook_rules = rules[rules['antecedents'].astype(str).str.contains('Asus EeePC')]

In [5]:
laptop_rules

antecedents  \
2                                    (ThinkPad X220 )   
13                (HP Laserjet P2055, ThinkPad X220 )   
14                (HP Laserjet P2055, ThinkPad X220 )   
15                    (HP CE50 Toner, ThinkPad X220 )   
17                    (HP CE50 Toner, ThinkPad X220 )   
42  (HP Laserjet P2055, HP CE50 Toner, ThinkPad X2...   
43  (HP Laserjet P2055, Lenovo Tablet Sleeve, Thin...   
44  (HP CE50 Toner, Lenovo Tablet Sleeve, ThinkPad...   
46                (HP Laserjet P2055, ThinkPad X220 )   
48                    (HP CE50 Toner, ThinkPad X220 )   

                                  consequents  antecedent support  \
2                      (Lenovo Tablet Sleeve)                 0.4   
13                     (Lenovo Tablet Sleeve)                 0.2   
14                            (HP CE50 Toner)                 0.2   
15                        (HP Laserjet P2055)                 0.2   
17                     (Lenovo Tablet Sleeve)                 0.2   
42                     (Lenovo Tablet Sleeve)                 0.2   
43                            (HP CE50 Toner)                 0.2   
44                        (HP Laserjet P2055)                 0.2   
46      (HP CE50 Toner, Lenovo Tablet Sleeve)                 0.2   
48  (HP Laserjet P2055, Lenovo Tablet Sleeve)                 0.2   

    consequent support  support  confidence      lift  leverage  conviction  \
2                  0.4      0.3        0.75  1.875000      0.14         2.4   
13                 0.4      0.2        1.00  2.500000      0.12         inf   
14                 0.3      0.2        1.00  3.333333      0.14         inf   
15                 0.3      0.2        1.00  3.333333      0.14         inf   
17                 0.4      0.2        1.00  2.500000      0.12         inf   
42                 0.4      0.2        1.00  2.500000      0.12         inf   
43                 0.3      0.2        1.00  3.333333      0.14         inf   
44                 0.3      0.2        1.00  3.333333      0.14         inf   
46                 0.2      0.2        1.00  5.000000      0.16         inf   
48                 0.2      0.2        1.00  5.000000      0.16         inf   

    zhangs_metric  
2        0.777778  
13       0.750000  
14       0.875000  
15       0.875000  
17       0.750000  
42       0.750000  
43       0.875000  
44       0.875000  
46       1.000000  
48       1.000000

In [6]:
printer_rules

antecedents  \
7                                 (HP Laserjet P2055)   
12          (HP Laserjet P2055, Lenovo Tablet Sleeve)   
13                (HP Laserjet P2055, ThinkPad X220 )   
14                (HP Laserjet P2055, ThinkPad X220 )   
33          (HP Laserjet P2055, Lenovo Tablet Sleeve)   
41  (HP Laserjet P2055, HP CE50 Toner, Lenovo Tabl...   
42  (HP Laserjet P2055, HP CE50 Toner, ThinkPad X2...   
43  (HP Laserjet P2055, Lenovo Tablet Sleeve, Thin...   
45          (HP Laserjet P2055, Lenovo Tablet Sleeve)   
46                (HP Laserjet P2055, ThinkPad X220 )   

                              consequents  antecedent support  \
7                         (HP CE50 Toner)                 0.3   
12                       (ThinkPad X220 )                 0.2   
13                 (Lenovo Tablet Sleeve)                 0.2   
14                        (HP CE50 Toner)                 0.2   
33                        (HP CE50 Toner)                 0.2   
41                       (ThinkPad X220 )                 0.2   
42                 (Lenovo Tablet Sleeve)                 0.2   
43                        (HP CE50 Toner)                 0.2   
45        (HP CE50 Toner, ThinkPad X220 )                 0.2   
46  (HP CE50 Toner, Lenovo Tablet Sleeve)                 0.2   

    consequent support  support  confidence      lift  leverage  conviction  \
7                  0.3      0.3         1.0  3.333333      0.21         inf   
12                 0.4      0.2         1.0  2.500000      0.12         inf   
13                 0.4      0.2         1.0  2.500000      0.12         inf   
14                 0.3      0.2         1.0  3.333333      0.14         inf   
33                 0.3      0.2         1.0  3.333333      0.14         inf   
41                 0.4      0.2         1.0  2.500000      0.12         inf   
42                 0.4      0.2         1.0  2.500000      0.12         inf   
43                 0.3      0.2         1.0  3.333333      0.14         inf   
45                 0.2      0.2         1.0  5.000000      0.16         inf   
46                 0.2      0.2         1.0  5.000000      0.16         inf   

    zhangs_metric  
7           1.000  
12          0.750  
13          0.750  
14          0.875  
33          0.875  
41          0.750  
42          0.750  
43          0.875  
45          1.000  
46          1.000

In [7]:
Netbook_rules

antecedents  \
3                                        (Asus EeePC)   
18                        (Asus EeePC, 2 GB DDR3 RAM)   
19                 (Asus EeePC, Netbook-Schutzhülle )   
23                        (Asus EeePC, LT Laser Maus)   
25                          (Asus EeePC, LT Minimaus)   
27                        (Asus EeePC, LT Laser Maus)   
29                          (Asus EeePC, LT Minimaus)   
30                 (Asus EeePC, Netbook-Schutzhülle )   
49         (Asus EeePC, LT Laser Maus, 2 GB DDR3 RAM)   
50  (Asus EeePC, LT Laser Maus, Netbook-Schutzhülle )   
52                        (Asus EeePC, LT Laser Maus)   
55           (Asus EeePC, LT Minimaus, 2 GB DDR3 RAM)   
56    (Asus EeePC, LT Minimaus, Netbook-Schutzhülle )   
57  (Asus EeePC, 2 GB DDR3 RAM, Netbook-Schutzhülle )   
59                          (Asus EeePC, LT Minimaus)   
60                 (Asus EeePC, Netbook-Schutzhülle )   

                              consequents  antecedent support  \
3                         (2 GB DDR3 RAM)                 0.6   
18                 (Netbook-Schutzhülle )                 0.5   
19                        (2 GB DDR3 RAM)                 0.4   
23                        (2 GB DDR3 RAM)                 0.2   
25                        (2 GB DDR3 RAM)                 0.4   
27                 (Netbook-Schutzhülle )                 0.2   
29                 (Netbook-Schutzhülle )                 0.4   
30                          (LT Minimaus)                 0.4   
49                 (Netbook-Schutzhülle )                 0.2   
50                        (2 GB DDR3 RAM)                 0.2   
52  (2 GB DDR3 RAM, Netbook-Schutzhülle )                 0.2   
55                 (Netbook-Schutzhülle )                 0.3   
56                        (2 GB DDR3 RAM)                 0.3   
57                          (LT Minimaus)                 0.4   
59  (2 GB DDR3 RAM, Netbook-Schutzhülle )                 0.4   
60           (LT Minimaus, 2 GB DDR3 RAM)                 0.4   

    consequent support  support  confidence      lift  leverage  conviction  \
3                  0.5      0.5    0.833333  1.666667      0.20         3.0   
18                 0.4      0.4    0.800000  2.000000      0.20         3.0   
19                 0.5      0.4    1.000000  2.000000      0.20         inf   
23                 0.5      0.2    1.000000  2.000000      0.10         inf   
25                 0.5      0.3    0.750000  1.500000      0.10         2.0   
27                 0.4      0.2    1.000000  2.500000      0.12         inf   
29                 0.4      0.3    0.750000  1.875000      0.14         2.4   
30                 0.5      0.3    0.750000  1.500000      0.10         2.0   
49                 0.4      0.2    1.000000  2.500000      0.12         inf   
50                 0.5      0.2    1.000000  2.000000      0.10         inf   
52                 0.4      0.2    1.000000  2.500000      0.12         inf   
55                 0.4      0.3    1.000000  2.500000      0.18         inf   
56                 0.5      0.3    1.000000  2.000000      0.15         inf   
57                 0.5      0.3    0.750000  1.500000      0.10         2.0   
59                 0.4      0.3    0.750000  1.875000      0.14         2.4   
60                 0.3      0.3    0.750000  2.500000      0.18         2.8   

    zhangs_metric  
3        1.000000  
18       1.000000  
19       0.833333  
23       0.625000  
25       0.555556  
27       0.750000  
29       0.777778  
30       0.555556  
49       0.750000  
50       0.625000  
52       0.750000  
55       0.857143  
56       0.714286  
57       0.555556  
59       0.777778  
60       1.000000

##### 8.1.3. Create association rules from the frequent item sets using the association_rules function with a confidence of 0.7. What do the rules tell you about the relationship between Asus EeePC netbooks, 2 GB DDR3 RAM extensions and Netbook Schutzhüllen? What do the lift values (compute it with the rules_stats function) tell you about the interestingness of the rules?

In [8]:
Netbook_rules[Netbook_rules['lift']>0.7]

antecedents  \
3                                        (Asus EeePC)   
18                        (Asus EeePC, 2 GB DDR3 RAM)   
19                 (Asus EeePC, Netbook-Schutzhülle )   
23                        (Asus EeePC, LT Laser Maus)   
25                          (Asus EeePC, LT Minimaus)   
27                        (Asus EeePC, LT Laser Maus)   
29                          (Asus EeePC, LT Minimaus)   
30                 (Asus EeePC, Netbook-Schutzhülle )   
49         (Asus EeePC, LT Laser Maus, 2 GB DDR3 RAM)   
50  (Asus EeePC, LT Laser Maus, Netbook-Schutzhülle )   
52                        (Asus EeePC, LT Laser Maus)   
55           (Asus EeePC, LT Minimaus, 2 GB DDR3 RAM)   
56    (Asus EeePC, LT Minimaus, Netbook-Schutzhülle )   
57  (Asus EeePC, 2 GB DDR3 RAM, Netbook-Schutzhülle )   
59                          (Asus EeePC, LT Minimaus)   
60                 (Asus EeePC, Netbook-Schutzhülle )   

                              consequents  antecedent support  \
3                         (2 GB DDR3 RAM)                 0.6   
18                 (Netbook-Schutzhülle )                 0.5   
19                        (2 GB DDR3 RAM)                 0.4   
23                        (2 GB DDR3 RAM)                 0.2   
25                        (2 GB DDR3 RAM)                 0.4   
27                 (Netbook-Schutzhülle )                 0.2   
29                 (Netbook-Schutzhülle )                 0.4   
30                          (LT Minimaus)                 0.4   
49                 (Netbook-Schutzhülle )                 0.2   
50                        (2 GB DDR3 RAM)                 0.2   
52  (2 GB DDR3 RAM, Netbook-Schutzhülle )                 0.2   
55                 (Netbook-Schutzhülle )                 0.3   
56                        (2 GB DDR3 RAM)                 0.3   
57                          (LT Minimaus)                 0.4   
59  (2 GB DDR3 RAM, Netbook-Schutzhülle )                 0.4   
60           (LT Minimaus, 2 GB DDR3 RAM)                 0.4   

    consequent support  support  confidence      lift  leverage  conviction  \
3                  0.5      0.5    0.833333  1.666667      0.20         3.0   
18                 0.4      0.4    0.800000  2.000000      0.20         3.0   
19                 0.5      0.4    1.000000  2.000000      0.20         inf   
23                 0.5      0.2    1.000000  2.000000      0.10         inf   
25                 0.5      0.3    0.750000  1.500000      0.10         2.0   
27                 0.4      0.2    1.000000  2.500000      0.12         inf   
29                 0.4      0.3    0.750000  1.875000      0.14         2.4   
30                 0.5      0.3    0.750000  1.500000      0.10         2.0   
49                 0.4      0.2    1.000000  2.500000      0.12         inf   
50                 0.5      0.2    1.000000  2.000000      0.10         inf   
52                 0.4      0.2    1.000000  2.500000      0.12         inf   
55                 0.4      0.3    1.000000  2.500000      0.18         inf   
56                 0.5      0.3    1.000000  2.000000      0.15         inf   
57                 0.5      0.3    0.750000  1.500000      0.10         2.0   
59                 0.4      0.3    0.750000  1.875000      0.14         2.4   
60                 0.3      0.3    0.750000  2.500000      0.18         2.8   

    zhangs_metric  
3        1.000000  
18       1.000000  
19       0.833333  
23       0.625000  
25       0.555556  
27       0.750000  
29       0.777778  
30       0.555556  
49       0.750000  
50       0.625000  
52       0.750000  
55       0.857143  
56       0.714286  
57       0.555556  
59       0.777778  
60       1.000000

1. The purchase of an Asus Eee PC is associated with buying 2 GB DDR3 RAM with a confidence level of 83%, demonstrating a positive interestingness of 1.67 (lift).
2. When both 2 GB DDR3 RAM and an Asus Eee PC are bought together, there's an 80% confidence that a Netbook-Schutzhülle (netbook case) will also be purchased, showing an interestingness of 2.0 (lift).

#### 8.2 Finding Frequent Pattern in the Adult Data Set

##### 8.2.1 Import the Adult-tweaked data set and print a description of the dataset using the describe function. The Adult-tweaked data set is provided on ILIAS as an ARFF file.

In [69]:
from scipy.io import arff
adult_arff_data, adult_arff_meta = arff.loadarff(open('adult-dataset-tweaked.arff', 'r'))
adult = pd.DataFrame(adult_arff_data)
adult = adult.applymap(lambda x: x.decode('utf8') if hasattr(x, 'decode') else x)
adult.head()

age  workclass    fnlwgt     education  education-num      marital-status  \
0  25.0    Private  226802.0          11th            7.0       Never-married   
1  25.0    Private  226802.0          11th            7.0       Never-married   
2  38.0    Private   89814.0       HS-grad            9.0  Married-civ-spouse   
3  28.0  Local-gov  336951.0    Assoc-acdm           12.0  Married-civ-spouse   
4  44.0    Private  160323.0  Some-college           10.0  Married-civ-spouse   

          occupation relationship   race   sex  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black     X           0.0           0.0   
1  Machine-op-inspct    Own-child  Black  Male           0.0           0.0   
2    Farming-fishing      Husband  White  Male           0.0           0.0   
3    Protective-serv      Husband  White  Male           0.0           0.0   
4  Machine-op-inspct      Husband  Black  Male        7688.0           0.0   

   hours-per-week native-country  class  
0            40.0  United-States      X  
1            40.0  United-States  <=50K  
2            50.0  United-States  <=50K  
3            40.0  United-States   >50K  
4            40.0  United-States   >50K

##### 8.2.2 Prepare the data set for Frequent Pattern Mining by: 1) reducing the size of the data set to 5000 examples using sampling; 2) using a subset of attributes: age, workclass, education, occupation, race, sex, hours-per-week, native-country and class; 3) discretizing the attributes age and hours-per-week into three user defined ranges (think about ranges that could make sense for the attributes). How many attributes does the resulting data set have?

In [76]:
# 1) reducing the size of the data set to 5000 examples using sampling;
sampled_df = adult.sample(n = 5000, random_state=42, axis = 0)

# 2) using a subset of attributes: age, workclass, education, occupation, race, sex, hours-per-week, native-country and class;

subset_df = sampled_df[['age','workclass','education','occupation','race','sex','hours-per-week','native-country','class']]
subset_df.head()

# 3) discretizing the attributes age and hours-per-week into three user defined ranges (think about ranges that could make sense for the attributes). 
# How many attributes does the resulting data set have?

import warnings
warnings.filterwarnings("ignore")

# from sklearn.preprocessing import KBinsDiscretizer
# import matplotlib.pyplot as plt
# encoder = KBinsDiscretizer()

# plt.figure(figsize = (10,3))

# for i in range(1,3):
#     axs = plt.subplot(1,2,i)
#     if i ==1:
#         axs.hist(subset_df[['age']])
#     elif i == 2:
#         axs.hist(subset_df[['hours-per-week']])
# plt.tight_layout()
# plt.show()


subset_df['age'] = pd.cut(subset_df['age'],[0,20,65,100],labels= ['low','middle','high'])
subset_df['hours-per-week'] = pd.cut(subset_df['hours-per-week'],[0,20,40,100],labels= ['part-time','full-time','over-time'])
display(subset_df.head())

# one hot encoder
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder(sparse= False)
oneHot_data = encoder.fit_transform(subset_df)
names = encoder.get_feature_names_out(subset_df.columns)
df = pd.DataFrame(oneHot_data,columns = names)
display(df)

age    workclass     education       occupation                race  \
7762   middle      Private     Bachelors  Exec-managerial               White   
23881  middle      Private       Masters  Exec-managerial               White   
30507  middle      Private       HS-grad     Adm-clerical               White   
28911  middle  Federal-gov  Some-college     Adm-clerical               White   
19484  middle      Private       HS-grad            Sales  Asian-Pac-Islander   

          sex hours-per-week native-country  class  
7762   Female      full-time  United-States  <=50K  
23881    Male      over-time  United-States   >50K  
30507    Male      over-time  United-States  <=50K  
28911    Male      full-time  United-States  <=50K  
19484    Male      over-time          South  <=50K

age_high  age_low  age_middle  workclass_?  workclass_Federal-gov  \
0          0.0      0.0         1.0          0.0                    0.0   
1          0.0      0.0         1.0          0.0                    0.0   
2          0.0      0.0         1.0          0.0                    0.0   
3          0.0      0.0         1.0          0.0                    1.0   
4          0.0      0.0         1.0          0.0                    0.0   
...        ...      ...         ...          ...                    ...   
4995       1.0      0.0         0.0          0.0                    0.0   
4996       0.0      0.0         1.0          0.0                    0.0   
4997       0.0      0.0         1.0          0.0                    0.0   
4998       0.0      0.0         1.0          0.0                    0.0   
4999       0.0      0.0         1.0          0.0                    0.0   

      workclass_Local-gov  workclass_Private  workclass_Self-emp-inc  \
0                     0.0                1.0                     0.0   
1                     0.0                1.0                     0.0   
2                     0.0                1.0                     0.0   
3                     0.0                0.0                     0.0   
4                     0.0                1.0                     0.0   
...                   ...                ...                     ...   
4995                  0.0                1.0                     0.0   
4996                  0.0                1.0                     0.0   
4997                  0.0                1.0                     0.0   
4998                  0.0                1.0                     0.0   
4999                  0.0                1.0                     0.0   

      workclass_Self-emp-not-inc  workclass_State-gov  ...  \
0                            0.0                  0.0  ...   
1                            0.0                  0.0  ...   
2                            0.0                  0.0  ...   
3                            0.0                  0.0  ...   
4                            0.0                  0.0  ...   
...                          ...                  ...  ...   
4995                         0.0                  0.0  ...   
4996                         0.0                  0.0  ...   
4997                         0.0                  0.0  ...   
4998                         0.0                  0.0  ...   
4999                         0.0                  0.0  ...   

      native-country_Puerto-Rico  native-country_Scotland  \
0                            0.0                      0.0   
1                            0.0                      0.0   
2                            0.0                      0.0   
3                            0.0                      0.0   
4                            0.0                      0.0   
...                          ...                      ...   
4995                         0.0                      0.0   
4996                         0.0                      0.0   
4997                         0.0                      0.0   
4998                         0.0                      0.0   
4999                         0.0                      0.0   

      native-country_South  native-country_Taiwan  native-country_Thailand  \
0                      0.0                    0.0                      0.0   
1                      0.0                    0.0                      0.0   
2                      0.0                    0.0                      0.0   
3                      0.0                    0.0                      0.0   
4                      1.0                    0.0                      0.0   
...                    ...                    ...                      ...   
4995                   0.0                    0.0                      0.0   
4996                   0.0                    0.0                      0.0   
4997                   0.0                    0.0           

##### 8.2.3 Find frequent item sets that have a support above 0.2. What can you learn from these item sets about the people who earn less than 50K a year?

In [102]:
from mlxtend.frequent_patterns import apriori

#replace column numbers with their names
frequent_patterns = apriori(df[df['class_<=50K'] == 1], min_support=0.2, use_colnames= True)
display(frequent_patterns)

support                                           itemsets
0    0.871019                                       (age_middle)
1    0.708081                                (workclass_Private)
2    0.345880                                (education_HS-grad)
3    0.246907                           (education_Some-college)
4    0.841800                                       (race_White)
..        ...                                                ...
218  0.301132  (workclass_Private, age_middle, class_<=50K, n...
219  0.314556  (workclass_Private, age_middle, class_<=50K, n...
220  0.219531  (workclass_Private, age_middle, class_<=50K, n...
221  0.259016  (age_middle, class_<=50K, native-country_Unite...
222  0.210582  (workclass_Private, class_<=50K, native-countr...

[223 rows x 2 columns]

##### 8.2.4 Given the large number of examples and the low min-support threshold, the number of frequent item sets containing the education attribute is surprisingly low. Moreover, only one value for this attribute is present in the resulting frequent item sets. Why is this the case? How could you aggregate the data to change this without losing too much information? Also look at the native-country attribute and think of a possible aggregation for it.

In [72]:
# binning the education column
subset_df['education'] = ['Basic education' if i in ['12th','HS-grad','11th','10th','7th-8th','9th','5th-6th','1st-4th','Preschool'] else i for i in subset_df['education']]
subset_df['education'] = ['vocational education' if i in ['Assoc-voc','Assoc-acdm','Prof-school','Some-college'] else i for i in subset_df['education']]
subset_df.loc[subset_df['native-country'] != 'United-States', 'native-country'] = 'Non-US'
display(subset_df.head())

# one hot encoder
encoder = OneHotEncoder(sparse= False)
oneHot_data = encoder.fit_transform(subset_df)
names = encoder.get_feature_names_out(subset_df.columns)
subset_df_encoded = pd.DataFrame(oneHot_data,columns = names)
display(subset_df_encoded)

age    workclass             education       occupation  \
7762   middle      Private             Bachelors  Exec-managerial   
23881  middle      Private               Masters  Exec-managerial   
30507  middle      Private       Basic education     Adm-clerical   
28911  middle  Federal-gov  vocational education     Adm-clerical   
19484  middle      Private       Basic education            Sales   

                     race     sex hours-per-week native-country  class  
7762                White  Female      full-time  United-States  <=50K  
23881               White    Male      over-time  United-States   >50K  
30507               White    Male      over-time  United-States  <=50K  
28911               White    Male      full-time  United-States  <=50K  
19484  Asian-Pac-Islander    Male      over-time         Non-US  <=50K

age_high  age_low  age_middle  workclass_?  workclass_Federal-gov  \
0          0.0      0.0         1.0          0.0                    0.0   
1          0.0      0.0         1.0          0.0                    0.0   
2          0.0      0.0         1.0          0.0                    0.0   
3          0.0      0.0         1.0          0.0                    1.0   
4          0.0      0.0         1.0          0.0                    0.0   
...        ...      ...         ...          ...                    ...   
4995       1.0      0.0         0.0          0.0                    0.0   
4996       0.0      0.0         1.0          0.0                    0.0   
4997       0.0      0.0         1.0          0.0                    0.0   
4998       0.0      0.0         1.0          0.0                    0.0   
4999       0.0      0.0         1.0          0.0                    0.0   

      workclass_Local-gov  workclass_Private  workclass_Self-emp-inc  \
0                     0.0                1.0                     0.0   
1                     0.0                1.0                     0.0   
2                     0.0                1.0                     0.0   
3                     0.0                0.0                     0.0   
4                     0.0                1.0                     0.0   
...                   ...                ...                     ...   
4995                  0.0                1.0                     0.0   
4996                  0.0                1.0                     0.0   
4997                  0.0                1.0                     0.0   
4998                  0.0                1.0                     0.0   
4999                  0.0                1.0                     0.0   

      workclass_Self-emp-not-inc  workclass_State-gov  ...  race_White  \
0                            0.0                  0.0  ...         1.0   
1                            0.0                  0.0  ...         1.0   
2                            0.0                  0.0  ...         1.0   
3                            0.0                  0.0  ...         1.0   
4                            0.0                  0.0  ...         0.0   
...                          ...                  ...  ...         ...   
4995                         0.0                  0.0  ...         1.0   
4996                         0.0                  0.0  ...         0.0   
4997                         0.0                  0.0  ...         1.0   
4998                         0.0                  0.0  ...         1.0   
4999                         0.0                  0.0  ...         0.0   

      sex_Female  sex_Male  hours-per-week_full-time  \
0            1.0       0.0                       1.0   
1            0.0       1.0                       0.0   
2            0.0       1.0                       0.0   
3            0.0       1.0                       1.0   
4            0.0       1.0                       0.0   
...          ...       ...                       ...   
4995         1.0       0.0                       1.0   
4996         0.0       1.0                       0.0   
4997         0.0       1.0                       0.0   
4998         1.0       0.0                       1.0   
4999         1.0       0.0                       1.0   

      hours-per-week_over-time  hours-per-week_part-time  \
0                          0.0                       0.0   
1                          1.0                       0.0   
2                          1.0                       0.0   
3                          0.0                       0.0   
4                          1.0                       0.0   
...                        ...                       ...   
4995                       0.0                       0.0   
4996                       0.0                       1.0   
4997                       1.0                       0.0   
4998                       0.0                       0.0   
4999              

##### 8.2.5 Restrict the frequent item sets to the ones containing “class = >50K” and lower the support so that a decent number of item sets is discovered. What can you learn from these item sets about the people who earn more than 50K a year?


In [75]:
#frequent item sets containing 'class_>50K

from mlxtend.frequent_patterns import apriori
column_mapping = {index:col_names for index,col_names in enumerate(subset_df_encoded.columns)} 
frequent_itemsets = apriori(subset_df_encoded, min_support=0.2)
frequent_itemsets['itemsets'] = frequent_itemsets['itemsets'].apply(lambda x: [column_mapping[item] for item in x])
display(frequent_itemsets[frequent_itemsets['itemsets'].astype(str).str.contains('class_>50K')])

support                                           itemsets
11    0.2402                                       [class_>50K]
22    0.2338                           [age_middle, class_>50K]
46    0.2194                           [race_White, class_>50K]
58    0.2196         [native-country_United-States, class_>50K]
81    0.2132               [age_middle, race_White, class_>50K]
92    0.2138  [age_middle, native-country_United-States, cla...
133   0.2068  [native-country_United-States, race_White, cla...
175   0.2012  [age_middle, race_White, native-country_United...

#### 8.3. Mining Association Rules from the Adult Data Set 

##### 8.3.1 Create association rules from the frequent item sets from exercise 8.2.3 using the association_rules function. Which rules do you consider interesting? Consider both =>50K and <50K classes.

In [93]:
# Association rule for 'class_<=50K'
from mlxtend.frequent_patterns import association_rules
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold= 0.9)
rules
rules[rules['consequents'].astype(str).str.contains('class_<=50K')]


antecedents    consequents  antecedent support  \
46  (sex_Female, hours-per-week_full-time)  (class_<=50K)               0.225   

    consequent support  support  confidence      lift  leverage  conviction  \
46              0.7598   0.2028    0.901333  1.186277  0.031845    2.434459   

    zhangs_metric  
46       0.202615

In [91]:
# Association rule for 'class_>50K'
from mlxtend.frequent_patterns import association_rules
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold= 0.25)
rules[rules['consequents'].astype(str).str.contains('class_>50K')]


antecedents  \
21                                         (age_middle)   
69                                         (race_White)   
223                            (race_White, age_middle)   
287          (native-country_United-States, age_middle)   
528          (race_White, native-country_United-States)   
1031  (race_White, native-country_United-States, age...   
1035         (race_White, native-country_United-States)   
1036                           (race_White, age_middle)   
1037         (native-country_United-States, age_middle)   

                                     consequents  antecedent support  \
21                                  (class_>50K)              0.8956   
69                                  (class_>50K)              0.8590   
223                                 (class_>50K)              0.7676   
287                                 (class_>50K)              0.8004   
528                                 (class_>50K)              0.7894   
1031                                (class_>50K)              0.7034   
1035                    (class_>50K, age_middle)              0.7894   
1036  (class_>50K, native-country_United-States)              0.7676   
1037                    (class_>50K, race_White)              0.8004   

      consequent support  support  confidence      lift  leverage  conviction  \
21                0.2402   0.2338    0.261054  1.086819  0.018677    1.028221   
69                0.2402   0.2194    0.255413  1.063336  0.013068    1.020432   
223               0.2402   0.2132    0.277749  1.156323  0.028822    1.051989   
287               0.2402   0.2138    0.267116  1.112058  0.021544    1.036727   
528               0.2402   0.2068    0.261971  1.090637  0.017186    1.029499   
1031              0.2402   0.2012    0.286039  1.190838  0.032243    1.064204   
1035              0.2338   0.2012    0.254877  1.090150  0.016638    1.028287   
1036              0.2196   0.2012    0.262116  1.193605  0.032635    1.057618   
1037              0.2194   0.2012    0.251374  1.145735  0.025592    1.042711   

      zhangs_metric  
21         0.765172  
69         0.422435  
223        0.581712  
287        0.504843  
528        0.394611  
1031       0.540307  
1035       0.392665  
1036       0.697943  
1037       0.637265

##### 8.3.2 Now we want to focus on the relationship of the occupation, education and income of immigrants: instead of sampling the data set, filter the data set: native-country != United-States. Which rules do you consider interesting?


In [101]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules


# droping the US residents from the data set
immigrant_df = subset_df_encoded.drop('native-country_United-States', axis = 1)
immigrant_df.head()

#column mapping
col_mapping =  {index:col_names for index,col_names in enumerate(immigrant_df)}
frequent_itemsets = apriori(immigrant_df, min_support= 0.2)
frequent_itemsets['itemsets'] = frequent_itemsets['itemsets'].apply(lambda x:[col_mapping[item] for item in x])
display(frequent_itemsets)

#Association rules 

rules = association_rules(frequent_itemsets, metric='confidence', min_threshold= 0.1)
display(rules[rules['consequents'].astype(str).str.contains('class_>50K')])


support                                           itemsets
0     0.8956                                       [age_middle]
1     0.6896                                [workclass_Private]
2     0.4448                        [education_Basic education]
3     0.3242                   [education_vocational education]
4     0.8590                                       [race_White]
..       ...                                                ...
120   0.2490  [class_<=50K, race_White, sex_Male, hours-per-...
121   0.2190  [age_middle, race_White, sex_Male, hours-per-w...
122   0.2566  [age_middle, race_White, sex_Male, workclass_P...
123   0.2702  [age_middle, race_White, hours-per-week_full-t...
124   0.2224  [age_middle, race_White, sex_Male, hours-per-w...

[125 rows x 2 columns]

antecedents               consequents  antecedent support  \
19               (age_middle)              (class_>50K)              0.8956   
59               (race_White)              (class_>50K)              0.8590   
182  (race_White, age_middle)              (class_>50K)              0.7676   
184              (race_White)  (class_>50K, age_middle)              0.8590   
185              (age_middle)  (class_>50K, race_White)              0.8956   

     consequent support  support  confidence      lift  leverage  conviction  \
19               0.2402   0.2338    0.261054  1.086819  0.018677    1.028221   
59               0.2402   0.2194    0.255413  1.063336  0.013068    1.020432   
182              0.2402   0.2132    0.277749  1.156323  0.028822    1.051989   
184              0.2338   0.2132    0.248196  1.061572  0.012366    1.019148   
185              0.2194   0.2132    0.238053  1.085017  0.016705    1.024480   

     zhangs_metric  
19        0.765172  
59        0.422435  
182       0.581712  
184       0.411354  
185       0.750530

#### ChatGPT Bonus Exercises

##### C.1. Discuss Measures and Quality of Rules 

Discuss Association Analysis (Association Rule Mining) with ChatGPT. How are the metrics 
support, confidence, and lift used to evaluate the quality of discovered association rules? 
Discuss the following three scenarios and identify the scenario that likely indicates an
interesting rule:
1. Low support and confidence as well as a lift close to 1.
2. High support and low confidence and a lift close to 1.
3. High support and confidence as well as a lift greater than 1.

Support, confidence, and lift are common metrics used in association rule mining to evaluate the quality and significance of discovered rules. 

1. **Low support and confidence with lift close to 1**: 
   - **Support**: Support measures the frequency of occurrence of an itemset in the dataset. Low support indicates that the rule is not frequently applicable to the dataset.
   - **Confidence**: Confidence measures the reliability of the rule. Low confidence indicates that the rule has a low predictive power.
   - **Lift**: Lift measures how much more often the antecedent and consequent of a rule occur together compared to what would be expected if they were independent. A lift close to 1 indicates that there is no significant association between the antecedent and consequent beyond what would be expected by chance.
   - Scenario 1 likely indicates an uninteresting rule because both support and confidence are low, meaning the rule is infrequent and not reliable. Additionally, the lift close to 1 suggests that there is no significant association between the antecedent and consequent beyond chance.

2. **High support and low confidence with lift close to 1**:
   - **Support**: High support indicates that the itemset occurs frequently in the dataset.
   - **Confidence**: Low confidence indicates that although the rule is frequently applicable, it doesn't always hold true.
   - **Lift**: Lift close to 1 suggests that there is no significant association between the antecedent and consequent beyond what would be expected by chance.
   - Scenario 2 may indicate an interesting rule because it has high support, meaning the rule is applicable to a significant portion of the dataset. However, the low confidence suggests that the rule may not always hold true. The lift close to 1 indicates that the association between the antecedent and consequent is not significant beyond what would be expected by chance. 

In summary, while scenario 1 indicates an uninteresting rule due to low support, low confidence, and lift close to 1, scenario 2 may indicate an interesting rule due to high support, low confidence, and lift close to 1. However, further analysis and domain knowledge are necessary to determine the actual significance and usefulness of the discovered rules.

##### C.2 Generate code to compare the Apriori algorithm and the FPgrowth algorithm

• In the lecture you learned about the Apriori algorithm for generating frequent itemsets. In 
the exercise we have used the FPgrowth algorithm instead to produce such itemsets. Have 
ChatGPT explain the difference between both algorithms to you. Ask ChatGPT for some code 
to produce frequent itemsets using the apriori algorithm with the orange3 library.

• Will both algorithms produce the same itemsets? Discuss this question with ChatGPT and ask 
for code that applies both algorithms using the orange3 library and subsequently compares 
the generated frequent item sets. Apply the code for one of the datasets from the exercise.

Apriori and FP-Growth are both algorithms used for frequent itemset mining in association rule learning, but they use different approaches to achieve this goal. Here's a comparison of the two algorithms:

1. **Apriori Algorithm**:
   - Apriori is a classical algorithm for frequent itemset generation.
   - It works by generating candidate itemsets of length \( k \) from itemsets of length \( k-1 \) using a level-wise approach.
   - It scans the database multiple times to find frequent itemsets.
   - It uses a candidate generation-and-test approach, which can be computationally expensive, especially for large datasets and itemsets.

2. **FP-Growth Algorithm**:
   - FP-Growth (Frequent Pattern Growth) is a more modern and efficient algorithm compared to Apriori.
   - It constructs a compact data structure called an FP-tree (Frequent Pattern tree) to represent the dataset.
   - FP-Growth mines frequent itemsets directly from the FP-tree without the need for candidate generation.
   - It requires only two passes over the dataset, making it much faster and more memory-efficient than Apriori, especially for large datasets.

Example:

Let's say we have a transaction dataset containing the following transactions:

Transaction 1: {bread, milk, eggs}
Transaction 2: {bread, butter}
Transaction 3: {milk, butter}
Transaction 4: {bread, milk}

Using the Apriori algorithm:

- Suppose we set a minimum support threshold of 50%.
- In the first pass, we identify frequent single items: {bread}, {milk}, {butter}.
- In the second pass, we generate candidate pairs and prune those below the minimum support threshold: {(bread, milk)}, {(bread, butter)}, {(milk, butter)}.
- We find that only {(bread, milk)} meets the minimum support threshold.

Using the FP-Growth algorithm:

- We construct an FP-tree from the transactions.
- The FP-tree for the given transactions looks like this:
  ```
       null (root)
       /   |   \
    bread milk butter
     /     \
  milk   butter
  ```
- From the FP-tree, we can directly identify frequent itemsets without generating candidates.
- Frequent itemsets found: {bread}, {milk}, {butter}, {(bread, milk)}.

While both algorithms produce the same result, FP-Growth typically outperforms Apriori in terms of efficiency, especially on large datasets, because it doesn't require candidate generation and multiple passes over the data.

##### C.3 Self Assessment

• Ask ChatGPT to create a pen and paper exercise by giving you some transaction items to 
practice to calculate the confidence of some association rules.

• Ask ChatGPT to generate some multiple-choice questions for graduate students related to 
association analysis. Include questions regarding pre-processing nominal and continuous 
features and shortcomings of the metrics support, confidence and lift that are used to assess 
the relevancy of association rules

**Pen and Paper Exercise for Calculating Confidence of Association Rules:**

Consider the following transaction items:

Transactions:
1. {bread, milk, eggs}
2. {bread, butter}
3. {milk, butter}
4. {bread, milk}

Now, let's practice calculating the confidence of association rules using these transactions.

1. Calculate the confidence for the association rule: {bread} → {milk}

2. Calculate the confidence for the association rule: {milk} → {bread}

3. Calculate the confidence for the association rule: {bread, butter} → {milk}

4. Calculate the confidence for the association rule: {milk, butter} → {bread}


**Multiple-Choice Questions for Graduate Students on Association Analysis:**

1. Which of the following steps is NOT typically involved in preprocessing nominal features for association analysis?
   a) One-hot encoding
   b) Label encoding
   c) Binning
   d) Normalization
   
2. What is a potential drawback of applying discretization to continuous features before association analysis?
   a) It increases the interpretability of association rules.
   b) It can result in loss of information.
   c) It speeds up the association rule mining process.
   d) It makes the dataset easier to visualize.

3. Which of the following is a limitation of the support metric in assessing the relevancy of association rules?
   a) It does not consider the size of the dataset.
   b) It does not account for the co-occurrence of items by chance.
   c) It is sensitive to changes in the dataset.
   d) It does not provide any information about the strength of the association.

4. A high lift value in association rule mining indicates:
   a) A strong association between the antecedent and consequent.
   b) A weak association between the antecedent and consequent.
   c) No association between the antecedent and consequent.
   d) None of the above.

5. Confidence measures:
   a) The reliability of the rule.
   b) The frequency of occurrence of an itemset.
   c) How much more often the antecedent and consequent occur together compared to what would be expected if they were independent.
   d) None of the above.

6. Which of the following is NOT a common algorithm for association rule mining?
   a) K-means
   b) Apriori
   c) FP-Growth
   d) Eclat

7. In association rule mining, pruning refers to:
   a) Removing infrequent itemsets from consideration.
   b) Removing frequent itemsets from consideration.
   c) Removing rules with low confidence.
   d) None of the above.